src/hardwareKeys.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * @fileoverview This module contains the template for hardware keys implementation.
+ * @see module:config/config
+ * @module hardwareKeys
+ */
+
+/* eslint-disable no-unused-vars */
+
+/**
+ * The abstract class with callbacks to implement for the on-hardware keys processing.
+ */
+export class HardwareKeys {
+  constructor() {
+    if (this.constructor === HardwareKeys) {
+      throw new Error('Abstract classes can\'t be instantiated.');
+    }
+  }
+
+  /**
+   * Return serial number of the device containing the private keys. Only the first 16 bytes will be used.
+   * @returns {Promise<Uint8Array>} The serial number, 16 bytes
+   * @async
+   */
+  async serial_number() {
+    throw new Error('Method must be implemented.');
+  }
+
+  /**
+   * Return the creation date of the keys
+   * @returns {Date} The keys creation date
+   */
+  date() {
+    throw new Error('Method must be implemented.');
+  }
+
+  /**
+   * Implement here all the required steps to initialize the plugin. This method should be called first
+   * before the actual use.
+   * @returns {Promise<void>}
+   * @async
+   */
+  async init() {
+    throw new Error('Method must be implemented.');
+  }
+
+  /**
+   * Generate ECDHE secret from private key and public part of ephemeral key
+   *
+   * @param {Object} obj - An object argument for destructuring
+   * @param {Curve} obj.curve - Elliptic curve object
+   * @param {Uint8Array} obj.V - Public part of ephemeral key
+   * @param {Uint8Array} obj.Q - Recipient public key
+   * @param {Uint8Array} obj.d - Recipient private key
+   * @returns {Promise<{secretKey: Uint8Array, sharedKey: Uint8Array}>}
+   * @async
+   */
+  async agree({ curve, V, Q, d }) {
```

Comment:
1. Makes sense. I did not want to introduce too much revolutionary changes, hence kept the same names and API, but I like the direction for the better naming. Will keep that in mind.
2. Noted about the the `curveName` preference. In the first iteration I was focused on the curves, hence this object leak. Certainly this would require something else for the RSA.

Renamed as requested, and removed the obsolete secret key field from the return type.